certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,215 @@
+function requireGuardiansLinkedListIntegrity() {
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+}
+
+// This integrity rule verifies that if the addGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the new guardian is added to the guardian list,
+// - and no other account (guardian or not) is affected.
+rule addGuardianWorksAsExpected(env e, address guardian, uint256 threshold, address otherAccount) {
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+
+    requireGuardiansLinkedListIntegrity();
+
+    require guardian != otherAccount;
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    currentContract.addGuardianWithThreshold(e, safeContract, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert currentContract.isGuardian(safeContract, guardian);
+    assert otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount + 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies that the guardian can always be added considering ideal conditions.
+// Some of the conditions are:
+// - No value should be sent with the transaction.
+// - The threshold should be greater than 0. We can't have guardian(s) with a threshold of 0.
+// - The guardian count should be less than the maximum value to prevent overflow.
+// - The guardian should not be values such as zero, sentinel, or the Safe contract itself.
+// - The guardian should not be an owner of the Safe contract at the time of addition.
+// - The guardian should not be already added as a guardian.
+// - The threshold should be less than or equal to the total number of guardians.
+// - Safe contract should be the sender of the transaction.
+rule guardianCanAlwaysBeAdded(env e, address guardian, uint256 threshold) {
+    requireSocialRecoveryModuleEnabled();
+
+    require e.msg.value == 0;
+    require threshold > 0;
+    require guardianStorageContract.entries[safeContract].count < max_uint256; // To prevent overflow (Realistically can't reach).
+    requireGuardiansLinkedListIntegrity();
+
+    require guardian != 0;
+    require guardian != SENTINEL();
+    require guardian != safeContract;
+
+    require !safeContract.isOwner(guardian);
+    require !currentContract.isGuardian(safeContract, guardian);
+
+    require threshold <= guardianStorageContract.countGuardians(safeContract);
+
+    require e.msg.sender == safeContract;
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted && currentContract.isGuardian(safeContract, guardian);
```

Comment:
We should also check that the count updates right?

Reply:
Done 👍🏾